### DISCUSSION TITLE
How to use dynamic framework in ane

### DISCUSSION THREAD

**Discussion by nbyusm** [airsdk/Adobe-Runtime-Support#853] [Q&A]
xcode：
***.framework embed&sign 

How to embed this framework in ane

I'm in a hurry!
please help me!

**Comment by ajwfrost** [comment]
Embedding the framework is one thing - using it would be another. Do you have an ANE already, that needs to use a framework? If so then it's just a case of adding the platform options file with a `packagedDependency` item.

In case it helps: https://community.adobe.com/t5/air/ane-with-ios-embedded-custom-framework/td-p/6906671

If you're needing to call into this framework from AIR/ActionScript code, then you need to put together the ANE to start with which will mean creating the AS3 wrapper and library and then creating the ANE, with the above packaged dependency so that it links properly.

**Comment by nbyusm** [comment]
@ajwfrost I already have ane。



//------------------
extension
-platform-ios.xml
-extension.xml
-AneTestLib.swc
-iPhone-ARM
--libly.a
--library.swf
--Frameworks
---LYSDK.framework
//--------------------------

<linkerOptions>
		<option>-ObjC</option>
		<option>-ios_version_min 8.0</option>
		<option>-framework CoreMotion</option>
		<option>-framework CoreGraphics</option>
		<option>-framework SystemConfiguration</option>
		<option>-framework CFNetwork</option>
		<option>-framework Security</option>
		<option>-framework CoreTelephony</option>
		<option>-framework CoreFoundation</option>
		<option>-lc++.1</option>
		<option>-lz.1</option>
		<option>-lsqlite3</option>
		<option>-lclang_rt.ios</option>
		<option>-rpath @executable_path/Frameworks</option>
    </linkerOptions>
    <packagedDependencies>
		<packagedDependency>Frameworks/LYSDK.framework</packagedDependency>
    </packagedDependencies>

It doesn't work


error：
xception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Description: DYLD, Library not loaded: @rpath/LYSDK.framework/LYSDK | Referenced from: /var/containers/Bundle/Application/6D41D94D-7B3E-4DDB-9706-E5AAD54E6C29/AneIOSTest.app/AneIOSTest | Reason: image not found
Triggered by Thread:  0

Thread 0 Crashed:
0   dyld                          	0x0000000101426418 __abort_with_payload + 8

**Comment by nbyusm** [comment]
@ajwfrost This framework is dynamic

**Comment by nbyusm** [comment]
@ajwfrost 
Execute command: file LYSDK

The result is:
LYSDK: Mach-O universal binary with 2 architectures: [arm_v7:Mach-O dynamically linked shared library arm_v7] [arm64]
LYSDK (for architecture armv7):	Mach-O dynamically linked shared library arm_v7
LYSDK (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64

**Comment by ajwfrost** [comment]
What's the contents of `platform-ios.xml`?

**Comment by marchbold** [comment]
You also need to make sure you include the dynamic frameworks in your air app package in the `Frameworks` directory at the root level. This is separate to what you package in your ANE.

**Reply by d-angelovski** [reply]
I am trying to build an ANE which has a framework from external SDK for printers (BrotherSDK), and I am adding the framework in ANE as packagedDependency, but app crashes on start probably because its not on root level. If I include the framework after making the release build (I have a Resign shell script, that basically takes a release build IPA file, unpacks it, and copy Frameworks/, then resign it and pack new ipa), like this it works.

But I don't have the possibility to Debug the app like this trough intelliJ, because debugging is not including the Frameworks in root level of the IPA, making my process of developing the SDK very hard (no Trace or debug messages). 

So my question is regards this process, you say "include the dynamic frameworks in air app package", is there some automated way to do this? I was trying to find online something, but closest I got is your comment.

**Reply by ajwfrost** [reply]
@d-angelovski as I understand it, the comment above re. adding dynamic frameworks at the root level, is for when you're adding these without them being in the ANE. If you're packaging the framework into the ANE with the `packagedDependencies` option, then the issue may either be about the file location, or about the code signature.. are you able to check the IPA that is produced when you're doing the debugging build, and unzip it to see where the files are (and whether they are properly code-signed: you can validate this on a mac using `codesign -d -vvv Payload/AppName.app` perhaps?)

thanks

**Reply by d-angelovski** [reply]
Hello @ajwfrost , I found a way to pack the framework files also. In IntelliJ, I am adding a Frameworks folder in the project which later I am including it here (see image)
![image](https://user-images.githubusercontent.com/23104582/184884570-d2e6017e-d25d-43b2-91af-bac23709dab6.png)

First time this .framework is added in the packaging of the ANE, then again needs to be added in the root of project, and included before packaging the IPA, like this the frameworks are available also while debugging, and no need to do manual resign (unpack ipa, put framework/s inside and resign and repack). Just adding it in the ANE is not working, at least for my case.

**Reply by marchbold** [reply]
Yeah with dynamic frameworks you have to package them in the ane then include them in a Frameworks folder at the root level. This is how we have always done it. There is no way to automatically include a dynamic framework in an ANE, (but you can in an AIR package ;) with `apm` it generates the Frameworks folder for you from the airpackage)

**Comment by nbyusm** [comment]
@ajwfrost 
```xml
<platform xmlns="http://ns.adobe.com/air/extension/33.1"> 
    <description>openane-ly</description> 
    <copyright>LY IOS ANE</copyright>
	<sdkVersion>0.2</sdkVersion>
    <linkerOptions>
		<option>-ObjC</option>
		<option>-ios_version_min 8.0</option>
		<option>-framework CoreMotion</option>
		<option>-framework CoreGraphics</option>
		<option>-framework SystemConfiguration</option>
		<option>-framework CFNetwork</option>
		<option>-framework Security</option>
		<option>-framework CoreTelephony</option>
		<option>-framework CoreFoundation</option>
		<option>-lc++.1</option>
		<option>-lz.1</option>
		<option>-lsqlite3</option>
		<option>-lclang_rt.ios</option>
		<option>-rpath @executable_path/Frameworks</option>
    </linkerOptions>
    <packagedDependencies>
		<packagedDependency>Frameworks/LYSDK.framework</packagedDependency>
    </packagedDependencies>
</platform>
```

**Comment by nbyusm** [comment]
@marchbold where? In Flash Builder?

**Reply by marchbold** [reply]
If that's the IDE you are using then yeah, at the root level, ie alongside your swf and other application content.

**Reply by nbyusm** [reply]
There was a mistake after that
WARNING: could not locate iTunesMetadata.plist in archive!
WARNING: could not locate Payload/AneIOSTest.app/SC_Info/AneIOSTest.sinf in archive!
ERROR: Install failed. Got error "ApplicationVerificationFailed" with code 0xe8008017: Failed to verify code signature of /private/var/installd/Library/Caches/com.apple.mobile.installd.staging/temp.evYj7d/extracted/Payload/AneIOSTest.app/Frameworks/LYSDK.framework : 0xe8008017 (A signed resource has been added, modified, or deleted.)

**Reply by marchbold** [reply]
Which version of AIR are you using? signing should be resolved in the latest build. 

Also make sure you don't have any files like `.DS_Store` in your framework folder.

**Reply by nbyusm** [reply]
The problem of resources in dynamic library。。。
But how to add the resources in the dynamic library?

**Reply by nbyusm** [reply]
LYSDK.framework
-Headers
-Modules
-SDKUI.bundle
-Info.plist
-LYSDK

**Comment by nbyusm** [comment]
@ajwfrost @marchbold The problem of resources in dynamic library。。。
